### PR TITLE
HDDS-9260. Optimize OmDbInsight API to avoid redundant recursive size calculations for deleted directories during data retrieval.

### PR DESCRIPTION
**Pull request by ArafatKhan2198**
## What changes were proposed in this pull request?

In the current implementation of `omdbInsights`, the calculation of sizes for `deleted directories` occurs dynamically when needed. The respective NSSummary object associated with a deleted directory does not retain the directory's size information. Consequently, we must repeatedly traverse the directory structure recursively to calculate its size whenever the omdbInsights endpoint is invoked.

To eliminate the redundancy, we can store the calculated size within the corresponding NSSummary objects. This means that the next time we need to determine the size of a specific deleted directory, we can retrieve the stored size value directly, avoiding the need for recomputation.

It's important to note that this approach is effective for **deleted directories**, as their size remains constant once they are deleted. Unlike normal directories, where size can change dynamically as files are added or removed, the size of deleted directories remains constant, making it a suitable candidate for this optimization.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9260

## How was this patch tested?
Integration Tests

### CODE REVIEW

**Comment by errose28** [comment]
cc @devmadhuu

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java [hunk 1]
```diff
@@ -580,10 +580,22 @@ protected long fetchSizeForDeletedDirectory(long objectId)
     if (nsSummary == null) {
       return 0L;
     }
+    // Check if the deleted dir size is already computed and set previously.
+    if (nsSummary.getIsSizeOfDeletedDirectoryComputed()) {
```
[the remark this answers is not in the document]
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java:584]: Thanks @devmadhuu I have added it along with Integration tests.

**Comment by ArafatKhan2198** [comment]
@devmadhuu @sumitagrawl @ashishkumar50 can you please look

**Review by devmadhuu — APPROVED** [review]
@ArafatKhan2198 thanks for working on this patch. LGTM +1

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/NSSummary.java [hunk 2]
```diff
@@ -82,6 +85,15 @@ public void setSizeOfFiles(long sizeOfFiles) {
     this.sizeOfFiles = sizeOfFiles;
   }
 
+  public boolean getIsSizeOfDeletedDirectoryComputed() {
+    return isSizeOfDeletedDirectoryComputed;
+  }
+
+  public void setIsSizeOfDeletedDirectoryComputed(
+      Boolean isSizeOfDeletedDirComputed) {
```
Comment (ashishkumar50) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/NSSummary.java:93]:
```suggestion
      boolean isSizeOfDeletedDirComputed) {
```

**Review by ashishkumar50 — COMMENTED** [review]
Thanks @ArafatKhan2198 for working on this.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/NSSummary.java (hunk 2 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/types/NSSummary.java]: Thanks I have made the changes

**Review by sumitagrawl — CHANGES_REQUESTED** [review]
@ArafatKhan2198 I have few scenario which seems not handled,
1. if sub-directory is removed, then the sub-dir size is not removed from parent -- This needs re-calculation again -- This is major issue with this implementation as needs recursive update till its parent
2. NSSummary when save to DB need extra field also, may not be much problem

**Comment by sumitagrawl** [comment]
Additionally,
- deleted directory size is not constant, it keeps changing as files and directory keeps deleting by directoryDeletingService.
- And NSSUmmary update is not saved to DB also while query update

**Pull request closed without merging**